feat(ctb): `OutputBisectionGame`

clabby opened the pull request:
## Overview

Implements the first iteration of the `OutputBisectionGame`, specified in [Notion](https://www.notion.so/oplabs/Revisiting-Output-Bisection-e76116f804fd428295ad5a501d74c786?pvs=4).

### Notes
- [x] Couple of follow-up nits from @tynes addressed.
- [x] Change local key context value.
- [x] Refactor unit tests that were disabled / add unit tests for changed functionality.
    - [x] Adding local inputs has changed a decent bit; Need new coverage for this.
    - [x] Claim verification on the split level - verification of the VM status byte has moved from when the contract is created to each time an exec trace subgame begins.
- [x] Handle attack on genesis in the output root bisection phase.
- [x] Protect against misconfigured depths (odd/even - who makes the step? who makes the exec trace root claims?).

**Metadata**
closes https://github.com/ethereum-optimism/client-pod/issues/94
closes https://github.com/ethereum-optimism/client-pod/issues/256

clabby committed:
Contract changes

clabby commented:
Current dependencies on/for this PR:
* develop
  * **PR #8351** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8351?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>  👈
    * **PR #8346** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8346?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8351?utm_source=stack-comment).

clabby committed:
Update local key loading + handle attack on genesis

clabby committed:
Remove `settlementHead` rename

clabby committed:
deploy config

clabby committed:
Add tests for invalid claims at the split depth

clabby committed:
Test cleanups

clabby committed:
:broom:

clabby committed:
Add split depth footgun checks

chores

ajsutton commented on op-chain-ops/genesis/config.go:
```diff
@@ -214,6 +214,8 @@ type DeployConfig struct {
 	// OutputBisectionGameGenesisBlock is the block number for genesis.
 	OutputBisectionGameGenesisBlock uint64 `json:"outputBisectionGameGenesisBlock"`
+	// OutputBisectionGameGenesisBlock is the block number for genesis.
+	OutputBisectionGameGenesisOutputRoot common.Hash `json:"outputBisectionGameGenesisOutputRoot"`
```
```suggestion
	// OutputBisectionGameGenesisBlock is the output root for the genesis block.
	OutputBisectionGameGenesisOutputRoot common.Hash `json:"outputBisectionGameGenesisOutputRoot"`
```

ajsutton commented on packages/contracts-bedrock/src/dispute/OutputBisectionGame.sol:
```diff
@@ -211,7 +218,7 @@ contract OutputBisectionGame is IOutputBisectionGame, Clone, ISemver {
         // When the next position surpasses the split depth (i.e., it is the root claim of an execution
         // trace bisection sub-game), we need to perform some extra verification steps.
-        if (nextPosition.depth() == SPLIT_DEPTH + 1) verifyExecBisectionRoot(_claim, _challengeIndex);
+        if (nextPosition.depth() == SPLIT_DEPTH + 1) verifyExecBisectionRoot(_claim);
```
Does our solidity style require braces on all if statements?

clabby replied:
Doesn't look like it. I normally omit if it doesn't surpass the line length and the body is a single statement.

ajsutton replied:
Yeah probably ok if it's on one line.  There's been some famous bugs because of the lack of braces when on separate line.

ajsutton commented on packages/contracts-bedrock/src/dispute/OutputBisectionGame.sol:
```diff
@@ -173,7 +180,7 @@ contract OutputBisectionGame is IOutputBisectionGame, Clone, ISemver {
         //            state's depth in relation to the parent, we don't need another
         //            branch because (n - n) % 2 == 0.
-        bool validStep = VM.step(_stateData, _proof, uuid) == Claim.unwrap(postState.claim);
+        bool validStep = VM.step(_stateData, _proof, Hash.unwrap(uuid)) == Claim.unwrap(postState.claim);
```
Will we need to translate the position to account for the split game before passing it to the VM or is that done elsewhere? Also doesn't need to be handled in this PR.

clabby replied:
We shouldn't - The search can't escape the execution trace bisection sub-tree:
![image](https://github.com/ethereum-optimism/optimism/assets/8406232/1c113dc8-b370-412d-ad20-e05db09b4e97)

ajsutton replied:
I was meaning that if the claim is the root of the execution trace that would be at `SPLIT_DEPTH` (maybe +1) but as far as the VM is concerned it's actually the root node of the tree.  So I was expecting that we'd need to translate the position to be relative to the execution trace root before passing it to the VM.

ajsutton replied:
I think you can ignore me on this.  We're not passing the claim position to the VM at all on the contract side because it's just given the preimage and we validate that only by checking that it's the pre-image.  We don't need to check that it's at the appropriate trace index.  Hence no need to translate the position like we need to on the challenger side.

clabby replied:
Ah great callout - yes, this should already be covered. The VM only cares about the UUID passed (which is a commitment to the starting/disputed claim values + positions, or only the disputed claim value + position in the case of the `GENESIS -> GENESIS + 1` trace), and this should be constant for all nodes within an instance of an execution trace subtree. As far as the native VM is concerned, there is only one set of local keys, but in the Solidity version, it's able to accept a `_localIdentifier` argument (the UUID) so that it can read from multiple sets of local keys belonging to the same `OutputBisectionGame`.

See: https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/cannon/MIPS.sol#L203-L208

ajsutton reviewed (commented):
Logic looks good to me. I'm guessing not all the bindings changes will be required when updated to the latest develop. I haven't tried to review the solidity tests carefully but all the solidity code should be reviewed by someone who knows what they're doing there anyway. :)

The local context calculation and finding the starting and disputed claims matches what the challenger does.

clabby committed:
Update op-chain-ops/genesis/config.go

Co-authored-by: Adrian Sutton <adrian@oplabs.co>

clabby committed:
Update op-chain-ops/genesis/config.go

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

tynes commented on packages/contracts-bedrock/snapshots/abi/OutputBisectionGame.json:
```diff
@@ -198,6 +203,19 @@
   },
+  {
+    "inputs": [],
+    "name": "GENESIS_OUTPUT_ROOT",
```
We are leaking the fact that its an immutable into the abi

clabby replied:
This is tracked in https://github.com/ethereum-optimism/client-pod/issues/258, and will be addressed lump-sum in another PR. There are also several other immutable vars in both the `OutputBisectionGame` + the `FaultDisputeGame` that don't follow the style guide standard.

tynes commented on packages/contracts-bedrock/src/dispute/OutputBisectionGame.sol:
```diff
@@ -173,7 +180,7 @@ contract OutputBisectionGame is IOutputBisectionGame, Clone, ISemver {
         //            state's depth in relation to the parent, we don't need another
         //            branch because (n - n) % 2 == 0.
-        bool validStep = VM.step(_stateData, _proof, uuid) == Claim.unwrap(postState.claim);
+        bool validStep = VM.step(_stateData, _proof, Hash.unwrap(uuid)) == Claim.unwrap(postState.claim);
```
Feels like the `Hash.unwrap` just creates unnecessary friction - should `VM.step` accept a `Hash`?

clabby replied:
Spoke offline - pushing this off due to the `MIPS` contract not respecting the dispute contracts' UDTs.

tynes commented on packages/contracts-bedrock/src/dispute/interfaces/IOutputBisectionGame.sol:
```diff
@@ -61,10 +61,7 @@ interface IOutputBisectionGame is IDisputeGame {
-    ///         first L2 Output Proposal in the `L2OutputOracle` that contains the disputed
-    ///         L2 block number.
+    /// @notice A block hash on the L1 that contains the disputed output root.
     function l1Head() external view returns (Hash l1Head_);
```
didn't you rename this?

clabby replied:
I did, although I wanted to double check with @ajsutton on the change due to some reliances within the `op-challenger`. This change is semantic and not critical for the functionality introduced in this PR.

refcell replied:
Can always change in a follow on pr?

tynes reviewed (commented):
I wouldn't say that I have the context to thoroughly review the implementation details but the code looks sane to me. Going to need to rely on testing for correctness

clabby commented on op-chain-ops/genesis/config.go:
```diff
@@ -214,6 +214,8 @@ type DeployConfig struct {
 	FaultGameMaxDuration uint64 `json:"faultGameMaxDuration"`
 	// OutputBisectionGameGenesisBlock is the block number for genesis.
 	OutputBisectionGameGenesisBlock uint64 `json:"outputBisectionGameGenesisBlock"`
+	// OutputBisectionGameGenesisOutputRoot is the output root for the genesis block.
+	OutputBisectionGameGenesisOutputRoot common.Hash `json:"outputBisectionGameGenesisOutputRoot"`
 	// OutputBisectionGameSplitDepth is the depth at which the output bisection game splits.
 	OutputBisectionGameSplitDepth uint64 `json:"outputBisectionGameSplitDepth"`
```
The `OutputBisectionGame...` types currently are devnet-specific, and excluded from the `Check()` function. We should circle back around on this once we remove the `onlyDevnet` limitation within the deploy script.

refcell commented on packages/contracts-bedrock/src/dispute/OutputBisectionGame.sol:
```diff
@@ -284,34 +291,34 @@ contract OutputBisectionGame is IOutputBisectionGame, Clone, ISemver {
-        // TODO(clabby): Include less context. See Adrian's proposal for the local context salt.
-        (ClaimData storage starting, ClaimData storage disputed) = findStartingAndDisputedOutputs(_execLeafIdx);
-        bytes32 uuid = keccak256(abi.encode(starting.claim, starting.parentIndex, disputed.claim, disputed.parentIndex));
+        (Claim starting, Position startingPos, Claim disputed, Position disputedPos) =
+            findStartingAndDisputedOutputs(_execLeafIdx);
+        Hash uuid = computeLocalContext(starting, startingPos, disputed, disputedPos);
```
Shouldn't we just be using `findLocalContext` here?
```suggestion
        Hash uuid = findLocalContext(_execLeafIdx);
```

clabby replied:
We unfortunately need access to the `starting` claim + position as well as the disputed claim below, so unfortunately not. Would be nice if we could, though

refcell replied:
MSTM

refcell reviewed (commented):
dogfood, otherwise lgtm

clabby committed:
Update `semver-lock.json`

refcell commented:
Need this in to fix tests in https://github.com/ethereum-optimism/optimism/pull/8394 since it relies on having the `GENESIS_OUTPUT_ROOT` exposed to validate the output bisection game top split

refcell commented:
![image](https://github.com/ethereum-optimism/optimism/assets/21288394/b51e4e8e-0e67-4720-9389-2f9c842bdf63)

ajsutton commented:
The test failing in the merge queue is for the output bisection game - exactly what this PR is changing.  I'd say it's legit broken when this PR combines with the latest changes on develop, but either way it definitely is getting flakier so we shouldn't just keep retrying the merge an need to work out how to fix it.

ajsutton committed:
Fix output_cannon e2e test.

Load the split depth from the contract when creating output_cannon games
Increase the max game depth so the cannon trace completes.
Legacy cannon games now have a fixed depth of 30 to avoid making those tests even slower.
Fix extra data creation in e2e helpers so the l2 block number is set correctly.

ajsutton commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -1027,7 +1027,7 @@ contract Deploy is Deployer {
             _gameType: GameTypes.FAULT,
             _absolutePrestate: loadMipsAbsolutePrestate(),
             _faultVm: IBigStepper(mustGetAddress("Mips")),
-            _maxGameDepth: cfg.faultGameMaxDepth()
+            _maxGameDepth: 30 // Hard code depth for legacy game to keep e2e tests fast
         });
     }
```
Have you considered that we made the change to deviate from the previous code deliberately?  And that maybe there's a comment explaining exactly why?

refcell commented:
> The test failing in the merge queue is for the output bisection game - exactly what this PR is changing. I'd say it's legit broken when this PR combines with the latest changes on develop, but either way it definitely is getting flakier so we shouldn't just keep retrying the merge an need to work out how to fix it.

The test failing is the codecov test which is saying 0% of the output bisection game is covered with tests (incorrect). This is why i was attempting to add to the merge queue a few times...

clabby commented on packages/contracts-bedrock/test/dispute/OutputBisectionGame.t.sol:
```diff
@@ -142,7 +197,7 @@ contract OutputBisectionGame_Test is OutputBisectionGame_Init {
         uint256 mask = 0xFF << offset;
         // Replace the byte in the slot value with the challenger wins status.
         slot = (slot & ~mask) | (chalWins << offset);
-        vm.store(address(gameProxy), bytes32(uint256(0)), bytes32(slot));
+        vm.store(address(gameProxy), bytes32(0), bytes32(slot));
```
Good callout, rabbit. Unfortunately, there's not a better way to do this without introspecting the storage layout in the contract artifact, and we don't anticipate this changing much.

ajsutton commented:
> The test failing is the codecov test which is saying 0% of the output bisection game is covered with tests (incorrect). This is why i was attempting to add to the merge queue a few times...

<img width="818" alt="image" src="https://github.com/ethereum-optimism/optimism/assets/72675/eb00ecb4-cb28-4d4f-ab77-f27b5965d1a9">

`TestOutputCannonGame` e2e test was failing and preventing the merge. The code coverage isn't actually a merge blocker. 😀
